juju/juju
# [JUJU-2741] Transition new Dqlite nodes into HA based on broadcast server details

This patch completes the ability to form Dqlite clusters based on addresses in the the Juju HA space if it is configured.

Hosts added to an existing HA cluster now set the appropriate bind address, and join the cluster based on API server details broadcast by the `peergrouper`.

We will now only rebind the original Dqlite node if there are multiple members known by the `peergrouper` worker. This fixes an issue on Microk8s introduced by the prior work.

Testing this patch revealed a race whereby:
- Newly added cluster nodes get their agent configs rewritten with `StateServingInfo` causing an agent restart *before* the machine agent has updated its status to `running`.
- Having `StateServingInfo` indicates a controller agent and means that the API caller only attempts to connect to localhost.
- But the `peergrouper` has not clustered the host (it remains in a pending state) so there is no API being served.
- The machine can now never set its status, and so is never joined to the cluster by the `peergrouper`.

The solution until we have an alternative is to allow controllers to connect to exiting API addresses in addition to localhost. This will be discussed within the team.

Logic relating to `TrackedDB` is moved into its own file to aid clarity in the worker module.

Changes to `NodeManager` methods `WithAddressOption` and `WithClusterOption` mean the removal of unused logic for bind address determination.

## QA steps

On LXD:
- Bootstrap.
- `juju enable-ha`.
- Once quiesced, check the logs for errors and blocked/restarting workers.

On Microk8s:
- `make microk8s-operator-update`.
- Bootstrap.
- Check that the logs show workers to be up and stable.
- Deploy a workload and ensure success.
